# Exploratory Data Analysis of House Price 
* Kaggle-Competition
    * House Prices: Advanced Regression Techniques
        * <a href='https://www.kaggle.com/c/house-prices-advanced-regression-techniques'>House Prices: Advanced Regression Techniques</a>
        * <a href='https://www.kaggle.com/c/house-prices-advanced-regression-techniques/data'>Data Description</a>

* Reference 
    * https://www.kaggle.com/siddheshpujari/exploratory-data-analysis-of-house-price

## Import Libraries

In [101]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
%matplotlib inline
import matplotlib.gridspec as gridspec
from matplotlib.gridspec import GridSpec

import seaborn as sns

# Plotly
import plotly.express as px
import plotly.graph_objs as go
import plotly.figure_factory as ff

# Some styling
sns.set_style("darkgrid")
plt.style.use("fivethirtyeight")

import plotly.io as pio
pio.templates.default = "gridon"

# Subplots
from plotly.subplots import make_subplots

# Showing full path of datasets

# Disable warnings
import warnings
warnings.filterwarnings('ignore')

In [102]:
df = pd.read_csv('./train.csv')
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities    ...     PoolArea PoolQC Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
1         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
2         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
3         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
4         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [103]:
df.tail()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities    ...     PoolArea PoolQC  Fence MiscFeature  \
1455         Lvl    AllPub    ...            0    NaN    NaN         NaN   
1456         Lvl    AllPub    ...            0    NaN  MnPrv         NaN   
1457         Lvl    AllPub    ...            0    NaN  GdPrv        Shed   
1458         Lvl    AllPub    ...            0    NaN    NaN         NaN   
1459         Lvl    AllPub    ...            0    NaN    NaN         NaN   

     MiscVal MoSold YrSold  SaleType  SaleCondition  SalePrice  
1455       0      8   2007        WD         Normal     175000  
1456       0      2   2010        WD         Normal     210000  
1457    2500      5   2010        WD         Normal     266500  
1458       0      4   2010        WD         Normal     142125  
1459       0      6   2008        WD         Normal     147500  

[5 rows x 81 columns]

In [104]:
# Number of rows and columns

df.shape

(1460, 81)

1460개의 row와 81개의 columns...!<br>
column의 수가 무려...81개다...!

In [105]:
# Columns in our dataset
df.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

In [106]:
# df.describe().T

In [107]:
# Description of our dataset

df.describe().T

# T refers to transpose that displays the description of our dataset in long format

count           mean           std      min        25%  \
Id             1460.0     730.500000    421.610009      1.0     365.75   
MSSubClass     1460.0      56.897260     42.300571     20.0      20.00   
LotFrontage    1201.0      70.049958     24.284752     21.0      59.00   
LotArea        1460.0   10516.828082   9981.264932   1300.0    7553.50   
OverallQual    1460.0       6.099315      1.382997      1.0       5.00   
OverallCond    1460.0       5.575342      1.112799      1.0       5.00   
YearBuilt      1460.0    1971.267808     30.202904   1872.0    1954.00   
YearRemodAdd   1460.0    1984.865753     20.645407   1950.0    1967.00   
MasVnrArea     1452.0     103.685262    181.066207      0.0       0.00   
BsmtFinSF1     1460.0     443.639726    456.098091      0.0       0.00   
BsmtFinSF2     1460.0      46.549315    161.319273      0.0       0.00   
BsmtUnfSF      1460.0     567.240411    441.866955      0.0     223.00   
TotalBsmtSF    1460.0    1057.429452    438.705324      0.0     795.75   
1stFlrSF       1460.0    1162.626712    386.587738    334.0     882.00   
2ndFlrSF       1460.0     346.992466    436.528436      0.0       0.00   
LowQualFinSF   1460.0       5.844521     48.623081      0.0       0.00   
GrLivArea      1460.0    1515.463699    525.480383    334.0    1129.50   
BsmtFullBath   1460.0       0.425342      0.518911      0.0       0.00   
BsmtHalfBath   1460.0       0.057534      0.238753      0.0       0.00   
FullBath       1460.0       1.565068      0.550916      0.0       1.00   
HalfBath       1460.0       0.382877      0.502885      0.0       0.00   
BedroomAbvGr   1460.0       2.866438      0.815778      0.0       2.00   
KitchenAbvGr   1460.0       1.046575      0.220338      0.0       1.00   
TotRmsAbvGrd   1460.0       6.517808      1.625393      2.0       5.00   
Fireplaces     1460.0       0.613014      0.644666      0.0       0.00   
GarageYrBlt    1379.0    1978.506164     24.689725   1900.0    1961.00   
GarageCars     1460.0       1.767123      0.747315      0.0       1.00   
GarageArea     1460.0     472.980137    213.804841      0.0     334.50   
WoodDeckSF     1460.0      94.244521    125.338794      0.0       0.00   
OpenPorchSF    1460.0      46.660274     66.256028      0.0       0.00   
EnclosedPorch  1460.0      21.954110     61.119149      0.0       0.00   
3SsnPorch      1460.0       3.409589     29.317331      0.0       0.00   
ScreenPorch    1460.0      15.060959     55.757415      0.0       0.00   
PoolArea       1460.0       2.758904     40.177307      0.0       0.00   
MiscVal        1460.0      43.489041    496.123024      0.0       0.00   
MoSold         1460.0       6.321918      2.703626      1.0       5.00   
YrSold         1460.0    2007.815753      1.328095   2006.0    2007.00   
SalePrice      1460.0  180921.195890  79442.502883  34900.0  129975.00   

                    50%        75%       max  
Id                730.5    1095.25    1460.0  
MSSubClass         50.0      70.00     190.0  
LotFrontage        69.0      80.00     313.0  
LotArea          9478.5   11601.50  215245.0  
OverallQual         6.0       7.00      10.0  
OverallCond         5.0       6.00       9.0  
YearBuilt        1973.0    2000.00    2010.0  
YearRemodAdd     1994.0    2004.00    2010.0  
MasVnrArea          0.0     166.00    1600.0  
BsmtFinSF1        383.5     712.25    5644.0  
BsmtFinSF2          0.0       0.00    1474.0  
BsmtUnfSF         477.5     808.00    2336.0  
TotalBsmtSF       991.5    1298.25    6110.0  
1stFlrSF         1087.0    1391.25    4692.0  
2ndFlrSF            0.0     728.00    2065.0  
LowQualFinSF        0.0       0.00     572.0  
GrLivArea        1464.0    1776.75    5642.0  
BsmtFullBath        0.0       1.00       3.0  
BsmtHalfBath        0.0       0.00       2.0  
FullBath            2.0       2.00       3.0  
HalfBath            0.0       1.00       2.0  
BedroomAbvGr        3.0       3.00       8.0  
KitchenAbvGr        1.0       1.00       3.0  
T

In [108]:
# Let's look at the skewness of our dataset

df.skew()

Id                0.000000
MSSubClass        1.407657
LotFrontage       2.163569
LotArea          12.207688
OverallQual       0.216944
OverallCond       0.693067
YearBuilt        -0.613461
YearRemodAdd     -0.503562
MasVnrArea        2.669084
BsmtFinSF1        1.685503
BsmtFinSF2        4.255261
BsmtUnfSF         0.920268
TotalBsmtSF       1.524255
1stFlrSF          1.376757
2ndFlrSF          0.813030
LowQualFinSF      9.011341
GrLivArea         1.366560
BsmtFullBath      0.596067
BsmtHalfBath      4.103403
FullBath          0.036562
HalfBath          0.675897
BedroomAbvGr      0.211790
KitchenAbvGr      4.488397
TotRmsAbvGrd      0.676341
Fireplaces        0.649565
GarageYrBlt      -0.649415
GarageCars       -0.342549
GarageArea        0.179981
WoodDeckSF        1.541376
OpenPorchSF       2.364342
EnclosedPorch     3.089872
3SsnPorch        10.304342
ScreenPorch       4.122214
PoolArea         14.828374
MiscVal          24.476794
MoSold            0.212053
YrSold            0.096269
S

*What is skeness?*

* Skewness tells us about the symmetry in a distribution
* If Skewness is equal to zero, It is a symmetrical distribution.
* And If <code>Skewness is less than or more than zero</code> then it is non-symmetrical distribution
* In our above data,
    * 1. LotArea
    * 2. LowQualFinSF
    * 3. 3SsnPorch
    * 4. PoolArea
    * 5. MiscVal
    
* Are highly positively, right skewed.

<code>pandas.DataFrame.info</code><br>
This method prints information about a DataFrame including the index dtype and column dtypes, non-null values and memory usage.

In [109]:
# Information of dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
Id               1460 non-null int64
MSSubClass       1460 non-null int64
MSZoning         1460 non-null object
LotFrontage      1201 non-null float64
LotArea          1460 non-null int64
Street           1460 non-null object
Alley            91 non-null object
LotShape         1460 non-null object
LandContour      1460 non-null object
Utilities        1460 non-null object
LotConfig        1460 non-null object
LandSlope        1460 non-null object
Neighborhood     1460 non-null object
Condition1       1460 non-null object
Condition2       1460 non-null object
BldgType         1460 non-null object
HouseStyle       1460 non-null object
OverallQual      1460 non-null int64
OverallCond      1460 non-null int64
YearBuilt        1460 non-null int64
YearRemodAdd     1460 non-null int64
RoofStyle        1460 non-null object
RoofMatl         1460 non-null object
Exterior1st      1460 non-n

The above information tells us

* Our dadtaset features consists of three datatypes
    * 1. float
    * 2. integer
    * 3. object
    
* Of which total **numerical features** are 38(floats + int64)
* And **categorical features** are 43
* But if we look closely, we see that some of the data types are incorrect.
    * *What?!*
* e.g - <code> MSSubClass</code>, <code>OverallQual</code>, <code>OverallCond</code> should be **object** data types.
* Also we don't have complete data for all of our features

In [110]:
df.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

* We can see a lot of columns in our dataset.
* We see that some of the features having int data type consists of discrete values.
* If the features have discrete values, it's better to change them into categorical variables that will help us in better anlysis.

✅ If the features have discrete values, it's better to change them into categorical variables that will help us in better analysis!

* **Note**
    * <code>df[column_name].astype()<code>

In [111]:
features_to_change = ['MSSubClass', 'OverallQual', 'OverallCond', 'YearBuilt', 
                     'MoSold', 'YrSold', 'GarageCars']  # ???
int_to_object = ['MSSubClass', 'OverallQual', 'OverallCond', 'GarageCars']
'''
for feature in int_to_object:
    df[feature] = df[feature].astype(object)  # convert int feature to obj feature
'''

for feature in int_to_object:
    df[feature] = df[feature].astype(object)  # convert int feature to obj feature!

int_to_datetime = ['YearBuilt', 'YearRemodAdd', 'MoSold', 'YrSold']

# We also have some datetime features, that we will deal later 
# For now keep them as int datatype


In [112]:
df.OverallCond.describe()

count     1460
unique       9
top          5
freq       821
Name: OverallCond, dtype: int64

It shows that 'OverallCond' feature changed to object datatype!

In [113]:
# Drop Id column as it is not required
# 필요없는 Id column은 drop시켜!

# df.drop(columns=['Id'], axis=1, inplace=True)  # axis=1:        column에서 drop
                                               # inplace=True:  실제 DataFrame 변경
df.head()
# df.drop(columns=['Id'], axis=1, inplace=True)

Id MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1         60       RL         65.0     8450   Pave   NaN      Reg   
1   2         20       RL         80.0     9600   Pave   NaN      Reg   
2   3         60       RL         68.0    11250   Pave   NaN      IR1   
3   4         70       RL         60.0     9550   Pave   NaN      IR1   
4   5         60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities    ...     PoolArea PoolQC Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
1         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
2         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
3         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
4         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   

  MoSold YrSold SaleType SaleCondition  SalePrice  
0      2   2008       WD        Normal     208500  
1      5   2007       WD        Normal     181500  
2      9   2008       WD        Normal     223500  
3      2   2006       WD       Abnorml     140000  
4     12   2008       WD        Normal     250000  

[5 rows x 81 columns]

🚦Do it!<br>
Drop Id feature from DataFrame

In [98]:
df.drop(labels='Id', axis=1).head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape LandContour  \
0         60       RL         65.0     8450   Pave   NaN      Reg         Lvl   
1         20       RL         80.0     9600   Pave   NaN      Reg         Lvl   
2         60       RL         68.0    11250   Pave   NaN      IR1         Lvl   
3         70       RL         60.0     9550   Pave   NaN      IR1         Lvl   
4         60       RL         84.0    14260   Pave   NaN      IR1         Lvl   

  Utilities LotConfig    ...     PoolArea PoolQC Fence MiscFeature MiscVal  \
0    AllPub    Inside    ...            0    NaN   NaN         NaN       0   
1    AllPub       FR2    ...            0    NaN   NaN         NaN       0   
2    AllPub    Inside    ...            0    NaN   NaN         NaN       0   
3    AllPub    Corner    ...            0    NaN   NaN         NaN       0   
4    AllPub       FR2    ...            0    NaN   NaN         NaN       0   

  MoSold YrSold SaleType  SaleCondition  SalePrice  
0      2   2008       WD         Normal     208500  
1      5   2007       WD         Normal     181500  
2      9   2008       WD         Normal     223500  
3      2   2006       WD        Abnorml     140000  
4     12   2008       WD         Normal     250000  

[5 rows x 80 columns]

In [114]:
df.drop(labels='Id', axis=1, inplace=True)
df.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape LandContour  \
0         60       RL         65.0     8450   Pave   NaN      Reg         Lvl   
1         20       RL         80.0     9600   Pave   NaN      Reg         Lvl   
2         60       RL         68.0    11250   Pave   NaN      IR1         Lvl   
3         70       RL         60.0     9550   Pave   NaN      IR1         Lvl   
4         60       RL         84.0    14260   Pave   NaN      IR1         Lvl   

  Utilities LotConfig    ...     PoolArea PoolQC Fence MiscFeature MiscVal  \
0    AllPub    Inside    ...            0    NaN   NaN         NaN       0   
1    AllPub       FR2    ...            0    NaN   NaN         NaN       0   
2    AllPub    Inside    ...            0    NaN   NaN         NaN       0   
3    AllPub    Corner    ...            0    NaN   NaN         NaN       0   
4    AllPub       FR2    ...            0    NaN   NaN         NaN       0   

  MoSold YrSold SaleType  SaleCondition  SalePrice  
0      2   2008       WD         Normal     208500  
1      5   2007       WD         Normal     181500  
2      9   2008       WD         Normal     223500  
3      2   2006       WD        Abnorml     140000  
4     12   2008       WD         Normal     250000  

[5 rows x 80 columns]

## Null Values

<code>df.isna()</code><br>
Detect missing values

In [115]:
# let's see if our dataset contains missing values

df.isna().sum().sum()

6965

6965 is a big number that tells us there are a lot of missing values.

In [116]:
type(df.isna().sum())

pandas.core.series.Series

In [117]:
df.isna().sum()

MSSubClass          0
MSZoning            0
LotFrontage       259
LotArea             0
Street              0
Alley            1369
LotShape            0
LandContour         0
Utilities           0
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
OverallQual         0
OverallCond         0
YearBuilt           0
YearRemodAdd        0
RoofStyle           0
RoofMatl            0
Exterior1st         0
Exterior2nd         0
MasVnrType          8
MasVnrArea          8
ExterQual           0
ExterCond           0
Foundation          0
BsmtQual           37
                 ... 
BedroomAbvGr        0
KitchenAbvGr        0
KitchenQual         0
TotRmsAbvGrd        0
Functional          0
Fireplaces          0
FireplaceQu       690
GarageType         81
GarageYrBlt        81
GarageFinish       81
GarageCars          0
GarageArea          0
GarageQual         81
GarageCond         81
PavedDrive

<code>df.isna().sum()</code> <- this doesn't help us much, let's try to visualize the number of missing values in each feature.

In [118]:
#First we create a list of missing values by each feature
temp = list(df.isna().sum())

#then we create a list of columns and their missing values as inner list to a separate list
lst= []
i=0
for col in df.columns:
    insert_lst = [col,temp[i]]
    lst.append(insert_lst)
    i+=1

#finally create a dataframe
temp_df = pd.DataFrame(data=lst,columns=['Column_Name','Missing_Values'])

In [119]:
fig = px.bar(temp_df.sort_values(by='Missing_Values'),x='Missing_Values',y='Column_Name',
             orientation='h',height=1500,width=900,color='Missing_Values',text='Missing_Values')
fig.update_traces(textposition='outside')
fig.show()

In [120]:
temp_df[temp_df['Missing_Values']>0].sort_values(by='Missing_Values',
                                                ascending=False).reset_index(drop=True).style.background_gradient(cmap='Reds')

*Q. 좀 더 쉽게 <code>pandas.Series</code>을 <code>DataFrame</code>로 바꿀 수는 없을까?*

* <code>pandas.Series.to_frame</code>
    * <a href='https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.to_frame.html'>pandas.Series.to_frame</a>
    * convert Series to DataFrame

In [122]:
df_isna = df.isna().sum().to_frame()
df_isna = df_isna.reset_index()

In [123]:
df_isna.columns

Index(['index', 0], dtype='object')

In [125]:
df_isna.rename(columns={"feature":"feature", 0: "null_num"}, inplace=True)
# df_isna

In [129]:
df_isna[df_isna.null_num>0].sort_values(by='null_num', ascending=False).style.background_gradient(cmap='Reds')

In [135]:
df_isna = df.isna().sum().to_frame().reset_index().rename(columns={'index':'class', 0:"num"})
df_isna

class   num
0      MSSubClass     0
1        MSZoning     0
2     LotFrontage   259
3         LotArea     0
4          Street     0
5           Alley  1369
6        LotShape     0
7     LandContour     0
8       Utilities     0
9       LotConfig     0
10      LandSlope     0
11   Neighborhood     0
12     Condition1     0
13     Condition2     0
14       BldgType     0
15     HouseStyle     0
16    OverallQual     0
17    OverallCond     0
18      YearBuilt     0
19   YearRemodAdd     0
20      RoofStyle     0
21       RoofMatl     0
22    Exterior1st     0
23    Exterior2nd     0
24     MasVnrType     8
25     MasVnrArea     8
26      ExterQual     0
27      ExterCond     0
28     Foundation     0
29       BsmtQual    37
..            ...   ...
50   BedroomAbvGr     0
51   KitchenAbvGr     0
52    KitchenQual     0
53   TotRmsAbvGrd     0
54     Functional     0
55     Fireplaces     0
56    FireplaceQu   690
57     GarageType    81
58    GarageYrBlt    81
59   GarageFinish    81
60     GarageCars     0
61     GarageArea     0
62     GarageQual    81
63     GarageCond    81
64     PavedDrive     0
65     WoodDeckSF     0
66    OpenPorchSF     0
67  EnclosedPorch     0
68      3SsnPorch     0
69    ScreenPorch     0
70       PoolArea     0
71         PoolQC  1453
72          Fence  1179
73    MiscFeature  1406
74        MiscVal     0
75         MoSold     0
76         YrSold     0
77       SaleType     0
78  SaleCondition     0
79      SalePrice     0

[80 rows x 2 columns]

In [143]:
df_isna[df_isna.num > 0].sort_values(by='num', ascending=False).reset_index().drop(labels='index', axis=1)

class   num
0         PoolQC  1453
1    MiscFeature  1406
2          Alley  1369
3          Fence  1179
4    FireplaceQu   690
5    LotFrontage   259
6     GarageType    81
7    GarageYrBlt    81
8   GarageFinish    81
9     GarageQual    81
10    GarageCond    81
11  BsmtExposure    38
12  BsmtFinType2    38
13  BsmtFinType1    37
14      BsmtCond    37
15      BsmtQual    37
16    MasVnrArea     8
17    MasVnrType     8
18    Electrical     1

In [144]:
df_isna = df_isna[df_isna.num > 0].sort_values(by='num', ascending=False).reset_index().drop(labels='index', axis=1)
df_isna

class   num
0         PoolQC  1453
1    MiscFeature  1406
2          Alley  1369
3          Fence  1179
4    FireplaceQu   690
5    LotFrontage   259
6     GarageType    81
7    GarageYrBlt    81
8   GarageFinish    81
9     GarageQual    81
10    GarageCond    81
11  BsmtExposure    38
12  BsmtFinType2    38
13  BsmtFinType1    37
14      BsmtCond    37
15      BsmtQual    37
16    MasVnrArea     8
17    MasVnrType     8
18    Electrical     1

In [146]:
df_isna.style.background_gradient(cmap='Reds')

In [157]:
df_class = df_isna['class'].tolist()
print(type(df_class))
print(df_class)

<class 'list'>
['PoolQC', 'MiscFeature', 'Alley', 'Fence', 'FireplaceQu', 'LotFrontage', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageQual', 'GarageCond', 'BsmtExposure', 'BsmtFinType2', 'BsmtFinType1', 'BsmtCond', 'BsmtQual', 'MasVnrArea', 'MasVnrType', 'Electrical']


In [167]:
# df.info()
non_obj = []

for feature in df_class:
#     print(feature, ":", df[feature].dtype)
    if df[feature].dtype != object:
        print(feature, ":", df[feature].dtype)
        non_obj.append(feature)

LotFrontage : float64
GarageYrBlt : float64
MasVnrArea : float64


* Out of the 18 columns with missing values
* There are numerical features <code>LostFrontage, MassVnrArea, GarageYrBlt</code>
* And the rest are categorical features

## Firstly, we'll deal with the numerical features. 

🏷 Numll Valeus -> deal with the numerical feature